backend/dal/Services/Concrete/BuildingService.cs

```diff
@@ -193,14 +227,132 @@ public Building Update(Building building)
             // Only administrators can dispose a property.
             if (building.ClassificationId == 4 && !isAdmin) throw new NotAuthorizedException("Building classification cannot be changed to disposed.");
 
+            var allowEdit = isAdmin || userAgencies.Contains(existingBuilding.AgencyId);
             // A building should have a unique name within the parcel it is located on.
             existingBuilding.Parcels.ForEach(pb => this.Context.ThrowIfNotUnique(pb.Parcel, building));
 
-            this.Context.Entry(existingBuilding).CurrentValues.SetValues(building);
+            foreach (var parcel in building.Parcels.Select(pb => pb.Parcel))
+            {
+                // Check if the building already exists.
+                var existingAssociatedParcel = existingBuilding.Parcels
+                    .FirstOrDefault(pb => pb.ParcelId == parcel.Id)?.Parcel;
+
+                // Reset all relationships that are not changed through this update.
+                parcel.Address.Province = this.Context.Provinces.FirstOrDefault(p => p.Id == parcel.Address.ProvinceId);
+                parcel.Classification = this.Context.PropertyClassifications.FirstOrDefault(b => b.Id == parcel.ClassificationId);
+                parcel.Agency = this.Context.Agencies.FirstOrDefault(a => a.Id == building.AgencyId);
+
+                if (existingAssociatedParcel == null)
+                {
+                    if (!allowEdit) throw new NotAuthorizedException("User may not add parcels to a property they don't own.");
+                    var existingParcel = this.Context.Parcels
+                    .FirstOrDefault(pb => pb.Id == parcel.Id);
+                    if(existingParcel != null)
+                    {
+                        _service.Parcel.PendingUpdate(parcel);
+                    }
+
+                    existingBuilding.Parcels.Add(new ParcelBuilding(existingParcel ?? parcel, building));
+                }
+                else
+                {
+                    _service.Parcel.PendingUpdate(parcel);
+                }
+            }
+
+            if (allowEdit)
+            {
+                this.Context.Entry(existingBuilding.Address).CurrentValues.SetValues(building.Address);
+                this.Context.Entry(existingBuilding).CurrentValues.SetValues(building);
+                this.Context.SetOriginalRowVersion(existingBuilding);
+            }
+
+            if (allowEdit)
```

Comment:
Any reason you can't include this if statement with the one above?

Reply:
Not that I can see.